File: frontend/src/services/employers.ts
import { apiFetch } from "./api";

export type Employer = {
  id: string;
  name: string;
  description: string | null;
  industry: string | null;
  website: string | null;
  location: string | null;
  contact_email?: string | null;
  status: string;
  created_at: string;
  updated_at: string;
};

export type Member = {
  id: string;
  employer_id: string;
  user_id: string;
  role: string;
  created_at: string;
};

export type Requirement = {
  id: string;
  employer_id: string;
  title: string;
  role_key: string | null;
  location: string | null;
  employment_type: string | null;
  description: string | null;
  experience_min_years: number | null;
  qualification_text: string | null;
  status: string;
  created_at: string;
  updated_at: string;
};

export type CanonicalSkill = {
  id: string;
  canonical_name: string;
  display_name: string;
  category?: string | null;
};

export type RequirementSkill = {
  id: string;
  hiring_requirement_id: string;
  skill_id: string;
  importance: "required" | "preferred";
  required_level: number | null;
  note: string | null;
  created_at: string;
  skill_name?: string | null;
  skill_category?: string | null;
};

export function listEmployers() {
  return apiFetch<Employer[]>("/employers");
}

export function createEmployer(payload: {
  name: string;
  description?: string | null;
  industry?: string | null;
  website?: string | null;
  location?: string | null;
  contact_email?: string | null;
}) {
  return apiFetch<Employer>("/employers", { method: "POST", body: JSON.stringify(payload) });
}

export function getEmployer(id: string) {
  return apiFetch<Employer>(`/employers/${id}`);
}

export function updateEmployer(
  employerId: string,
  payload: {
    name?: string;
    description?: string | null;
    industry?: string | null;
    website?: string | null;
    location?: string | null;
    contact_email?: string | null;
    status?: "active" | "suspended" | "archived";
  }
) {
  return apiFetch<Employer>(`/employers/${employerId}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export function deleteEmployer(employerId: string) {
  return apiFetch<void>(`/employers/${employerId}`, { method: "DELETE" });
}

export function listMembers(employerId: string) {
  return apiFetch<Member[]>(`/employers/${employerId}/members`);
}

export function addMember(employerId: string, payload: { user_id: string; role?: "owner" | "member" }) {
  return apiFetch<Member>(`/employers/${employerId}/members`, {
    method: "POST",
    body: JSON.stringify({ role: "member", ...payload }),
  });
}

export function removeMember(employerId: string, targetUserId: string) {
  return apiFetch<void>(`/employers/${employerId}/members/${targetUserId}`, {
    method: "DELETE",
  });
}

export function listRequirements(employerId: string) {
  return apiFetch<Requirement[]>(`/employers/${employerId}/requirements`);
}

export function getRequirement(requirementId: string) {
  return apiFetch<Requirement>(`/requirements/${requirementId}`);
}

export function createRequirement(
  employerId: string,
  payload: {
    title: string;
    role_key?: string | null;
    location?: string | null;
    employment_type?: string | null;
    description?: string | null;
    experience_min_years?: number | null;
    qualification_text?: string | null;
  }
) {
  return apiFetch<Requirement>(`/employers/${employerId}/requirements`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateRequirement(
  requirementId: string,
  payload: {
    title?: string;
    role_key?: string | null;
    location?: string | null;
    employment_type?: string | null;
    description?: string | null;
    experience_min_years?: number | null;
    qualification_text?: string | null;
    status?: "draft" | "open" | "paused" | "closed";
  }
) {
  return apiFetch<Requirement>(`/requirements/${requirementId}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export function deleteRequirement(requirementId: string) {
  return apiFetch<void>(`/requirements/${requirementId}`, { method: "DELETE" });
}

export function listRequirementSkills(requirementId: string) {
  return apiFetch<RequirementSkill[]>(`/requirements/${requirementId}/skills`);
}

export function setRequirementSkills(
  requirementId: string,
  skills: { skill_id: string; importance: "required" | "preferred"; required_level?: number | null; note?: string | null }[]
) {
  return apiFetch<RequirementSkill[]>(`/requirements/${requirementId}/skills`, {
    method: "PUT",
    body: JSON.stringify({ skills }),
  });
}

export function listCanonicalSkills() {
  return apiFetch<CanonicalSkill[]>("/employers/skills/catalog");
}

File: backend/app/schemas/employers.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime
from uuid import UUID


class EmployerCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=200)
    description: Optional[str] = None
    industry: Optional[str] = Field(None, max_length=150)
    website: Optional[str] = Field(None, max_length=300)
    location: Optional[str] = Field(None, max_length=200)
    contact_email: Optional[str] = Field(None, max_length=200)


class EmployerUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=2, max_length=200)
    description: Optional[str] = None
    industry: Optional[str] = Field(None, max_length=150)
    website: Optional[str] = Field(None, max_length=300)
    location: Optional[str] = Field(None, max_length=200)
    contact_email: Optional[str] = Field(None, max_length=200)
    status: Optional[str] = Field(None, pattern="^(active|suspended|archived)$")


class EmployerOut(BaseModel):
    id: UUID
    name: str
    description: Optional[str] = None
    industry: Optional[str] = None
    website: Optional[str] = None
    location: Optional[str] = None
    contact_email: Optional[str] = None
    status: str
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class MemberOut(BaseModel):
    id: UUID
    employer_id: UUID
    user_id: str
    role: str
    created_at: datetime

    class Config:
        from_attributes = True


class MemberAdd(BaseModel):
    user_id: str = Field(..., min_length=1)
    role: str = Field("member", pattern="^(owner|member)$")


class RequirementCreate(BaseModel):
    title: str = Field(..., min_length=2, max_length=200)
    role_key: Optional[str] = Field(None, max_length=150)
    location: Optional[str] = Field(None, max_length=200)
    employment_type: Optional[str] = Field(
        None, pattern="^(full_time|part_time|internship|contract|apprenticeship)$"
    )
    description: Optional[str] = None
    experience_min_years: Optional[float] = Field(None, ge=0)
    qualification_text: Optional[str] = None


class RequirementUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=2, max_length=200)
    role_key: Optional[str] = Field(None, max_length=150)
    location: Optional[str] = Field(None, max_length=200)
    employment_type: Optional[str] = Field(
        None, pattern="^(full_time|part_time|internship|contract|apprenticeship)$"
    )
    description: Optional[str] = None
    experience_min_years: Optional[float] = Field(None, ge=0)
    qualification_text: Optional[str] = None
    status: Optional[str] = Field(None, pattern="^(draft|open|paused|closed)$")


class RequirementOut(BaseModel):
    id: UUID
    employer_id: UUID
    title: str
    role_key: Optional[str] = None
    location: Optional[str] = None
    employment_type: Optional[str] = None
    description: Optional[str] = None
    experience_min_years: Optional[float] = None
    qualification_text: Optional[str] = None
    status: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class RequirementSkillIn(BaseModel):
    skill_id: UUID
    importance: str = Field(..., pattern="^(required|preferred)$")
    required_level: Optional[float] = Field(None, ge=0, le=1)
    note: Optional[str] = None


class RequirementSkillOut(BaseModel):
    id: UUID
    hiring_requirement_id: UUID
    skill_id: UUID
    importance: str
    required_level: Optional[float] = None
    note: Optional[str] = None
    created_at: datetime
    skill_name: Optional[str] = None
    skill_category: Optional[str] = None

    class Config:
        from_attributes = True


class RequirementSkillsBulk(BaseModel):
    skills: List[RequirementSkillIn] = Field(..., min_length=0, max_length=50)


class CanonicalSkillOut(BaseModel):
    id: UUID
    canonical_name: str
    display_name: str
    category: Optional[str] = None

    class Config:
        from_attributes = True

